enhancements/ingress/optional-ingress.md

```diff
@@ -0,0 +1,225 @@
+
+However the implementation of these APIs (or parts of them) depend on the ingress operator:
+- The ingress operator creates the default ingress controller to manage the backends for the routes.
+- For [the component routes](https://docs.openshift.com/container-platform/4.13/rest_api/config_apis/ingress-config-openshift-io-v1.html#spec-componentroutes) the ingress operator creates the required role and role bindings to get the custom TLS certificate secret from `openshift-config` namespace.
+
+#### Console operator
+
+To prevent the console operator from going degraded, [the console route check](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/healthcheck/controller.go#L122) should be skipped if the ingress capability is not enabled in `ClusterVersion`.
+
+#### Authentication operator
+
+No change to the authentication operator is needed until the implementation of the ingress capability for the standalone OpenShift.
+HyperShift doesn't use the authentication operator.
+
+#### Monitoring operator
+
+No change to the monitoring operator is needed until the implementation of the ingress capability for the standalone OpenShift.
+The monitoring operator skips the alert manager's route check if `ingress` cluster operator is not there.
+
+### Risks and Mitigations
+
+#### Partially implemented capability
+
+The ingress capability will become a part of the capability API. Meanwhile its implementation will be incomplete due to the missing support of the standalone OpenShift. However the usage of the capability is an important stepping stone, it unifies the implementation of the optional ingress for all the OpenShift installation types.
+
+#### Component routes
+
+In the absence of the ingress operator, the component routes cannot be customized by the owner components due to the lack of the RBAC.
+The cluster administrator can customize a component route with a custom TLS certificate secret but the component owner won't be able to access the secret. This may be misleading for the cluster administrators. However the component routes were never fully implemented on HyperShift (authentication operator is not used).
+
+### Drawbacks
+
+In the absence of the default ingress controller, the Route API is left unserved. The OpenShift components or workloads can create routes but no controller is available to back them up. This may be misleading for the users.
+
+## Design Details
+
+### Open Questions [optional]
+
+N/A
+
+### Test Plan
+TBD
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+N/A
+
+#### Tech Preview -> GA
+N/A
+
+#### Removing a deprecated feature
+N/A
+
+### Upgrade / Downgrade Strategy
+
+Same as the component selection [strategy](/enhancements/installer/component-selection.md#upgrade--downgrade-strategy).
```

Comment:
That doc says on upgrades the CVO applies "the same exclusion filters". Does that mean that for a cluster that was deployed with baselineCapabilitySet=None, this new capability would not be updated and the old implementation would be present?

Reply:
Good question. Let me try to upgrade from one `None` cluster to another.

Reply:
I installed a 4.12.19 cluster with the following capabilities:
```yaml
capabilities:
  baselineCapabilitySet: None
  additionalEnabledCapabilities:
  - Console
```
That resulted in the same cluster version capabilities:
```yaml
spec:
  capabilities:
    additionalEnabledCapabilities:
      - Console
    baselineCapabilitySet: None
  channel: stable-4.12
  ...
status:
  capabilities:
    enabledCapabilities:
      - Console
    knownCapabilities:
      - CSISnapshot
      - Console
      - Insights
      - Storage
      - baremetal
      - marketplace
      - openshift-samples
```
Then I upgraded to 4.13.2 which has 1 known capability more: `NodeTuning`. The node-tuning operator was up and running on the version 4.12.19. After the upgrade started the cluster version has been updated to this state:
```yaml
spec:
  capabilities:
    additionalEnabledCapabilities:
      - Console
    baselineCapabilitySet: None
  channel: stable-4.13
  ...
status:
  capabilities:
    enabledCapabilities:
      - Console
      - NodeTuning
    knownCapabilities:
      - CSISnapshot
      - Console
      - Insights
      - NodeTuning
      - Storage
      - baremetal
      - marketplace
      - openshift-samples
```
And the following condition has been added to the cluster version:
```
ImplicitlyEnabledCapabilities
True
Jun 5, 2023, 2:59 PM
CapabilitiesImplicitlyEnabled
The following capabilities could not be disabled: NodeTuning
```

Maybe that what's meant by "the same exclusion filters" - no new capabilities will be excluded while upgrading to a new cluster version.

Reply:
I vaguely remember that something (CVO?) would recognize that manifests related to running components were annotated with a new capability and enable that capability. If that's the case, adding this new one should be OK, as existing clusters would keep running ingress and managing its manifests.

Maybe @bparees can help make sure we understand how that actually works.

Reply:
> I vaguely remember that something (CVO?) would recognize that manifests related to running components were annotated with a new capability and enable that capability.

yes that was part of the design.  @wking can confirm if it got implemented but i believe it did.  That's what implicitly enabled capabilities represent, primarily.

Reply:
I added a couple of notes about the implicit enabling and the forbidden disabling post install.

Reply:
Yes, we [did implement][1] implicitly enabled capabilities.

[1]: https://github.com/openshift/cluster-version-operator/blob/f74d114b1b2843f045fdcdc63ca63fa4cf07da75/pkg/payload/payload.go#L237-L240